src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+import json
+import re
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, field, fields, replace
+from datetime import datetime, timedelta, timezone
+from enum import IntEnum, auto, unique
+from functools import cached_property, lru_cache
+from itertools import chain, groupby, repeat, starmap
+from random import random
+from typing import (
+    Any,
+    BinaryIO,
+    Callable,
+    Dict,
+    Hashable,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    get_args,
+    overload,
+)
+from uuid import uuid4
+from weakref import ProxyType, proxy
+
+import numpy as np
+import pandas as pd
+from pandas.core.dtypes.common import (
+    is_datetime64_any_dtype,
+    is_datetime64tz_dtype,
+    is_numeric_dtype,
+)
+from typing_extensions import TypeAlias, TypeGuard
+from wrapt import ObjectProxy
+
+
+class DimensionRole(IntEnum):
+    ...
+
+
+@unique
+class SingularDimensionalRole(DimensionRole):
+    UNASSIGNED = auto()
+    PREDICTION_ID = auto()
+    TIME = auto()
+    PREDICTION_LABEL = auto()
+    PREDICTION_SCORE = auto()
+    ACTUAL_LABEL = auto()
+    ACTUAL_SCORE = auto()
+
+
+@unique
+class MultiDimensionalRole(DimensionRole):
```

Comment:
The naming here is a tad non-intuitive for me from a domain language perspective - Where the "singular" - are basically columns with reserved meaning, "multiple" is actually more of a "category" that can map to multiple columns. I can see a future where "predictions" is more of a category.

Not blocking but I would consider thinking in this terms where columns / dimensions can have a particular significance but also has a semantic grouping / category - which maps closer to roles.

Reply:
i didn't want to allude to the idea of reserve dimensions because i have tried to avoid it altogether by creating a distinction between the (true) name of a dimension (which is the name of the actual column in dataframe that's kept intact for the purpose of exports) and the display name, which in the case of singular dimensions, can be used as a alias (because the role has only one opening)